packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1071,11 +1071,18 @@ async function renderToHTMLOrFlightImpl(
     formState,
   })
 
+  const requestStoreWaitUntilTasks = requestStore.waitUntil.map((p) =>
+    typeof p === 'function' ? p() : p
+  )
+
   // If we have pending revalidates, wait until they are all resolved.
   if (staticGenerationStore.pendingRevalidates) {
-    options.waitUntil = Promise.all(
-      Object.values(staticGenerationStore.pendingRevalidates)
-    )
+    options.waitUntil = Promise.all([
```

Comment:
Like @dvoytenko mentioned above I think we can make this a bit smarter because waiting for `Promise.all` would not allow to keep extending the execution from an scheduled task. For the particular case of `after` I'm thinking of:

```typescript
after(async () => {
  await sleep(1000);
  after(() => {
    await sleep(1000);
    console.log('All done');
  });
})
```

Maybe we don't want to allow this from the `after` primitive but it is definitely possible to do this infra-wise.